Add Parent Queue Quota Checking to Scheduler.

shchennu opened the pull request:
# Add Parent Queue Quota and Limit Checking to KAI Scheduler

## Description
This PR enhances the KAI Scheduler with comprehensive parent queue resource management, implementing both quota and limit checking at parent queue levels. The changes improve scheduler efficiency by preventing unnecessary scheduling attempts and provide better resource management for different job types.

## Changes
- Added `checkParentQueueLimits` method to `CapacityPolicy` to check both limits and quotas at parent queue levels
- Added `getFirstPendingPod` helper function to identify the first pending pod in a job
- Modified `OnSessionOpen` to include early validation of resource limits
- Added comprehensive test cases for parent queue limit checking
- Added support for elastic jobs with minimum required resources
- Separated job priority from preemptibility handling

## Implementation Details
- Resource checks are performed for GPU, CPU, and Memory resources
- Both limits and quotas are checked, using the more restrictive value
- Checks are done at each parent queue level in the hierarchy
- Early validation prevents unnecessary scheduling attempts
- Proper error messages indicate which resource limit/quota was exceeded
- Special handling for preemptible jobs (PriorityInferenceNumber)
- Support for elastic jobs with minimum resource requirements

## Test Coverage
- Basic limit/quota checks for each resource type (GPU, CPU, Memory)
- Multi-level queue hierarchy checks
- Elastic job handling with minimum required resources
- Preemptible vs non-preemptible job behavior
- Edge cases (zero limits/quotas, missing values)
- Error message formatting
- Helper function validation

## Files Changed
- Modified: `pkg/scheduler/plugins/proportion/capacity_policy/capacity_policy.go`
- Modified: `pkg/scheduler/plugins/proportion/capacity_policy/parent_queue_test.go`

## Testing
All tests are passing:
- 37 specs from the original test suite
- 10 new test cases for parent queue limit checking
- 3 test cases for the helper function

## Impact
- Improves scheduler efficiency by preventing unnecessary scheduling attempts
- Provides clear error messages when limits/quotas are exceeded
- Better resource management for different job types (preemptible, elastic)
```

Commit:
Add Parent Queue Quota Checking to Scheduler.

romanbaron reviewed (commented):
Hi @shchennu 👋 
Thanks for suggesting this change, I want to better understand the motivation here, more specifically what is the gap with `IsJobOverQueueCapacity`? From what I see it is called before `PrePredicateFn` so there is probably something that you identified as a gap in the implementation?

Another question, what is the problem with job allocation above parent's queue quota? If the job is preemptible it can be allowed.

Commit:
Apply feedback priorityTrainNumber can exceed parent queue GPU quotas

shchennu commented:
> Hi @shchennu 👋 Thanks for suggesting this change, I want to better understand the motivation here, more specifically what is the gap with `IsJobOverQueueCapacity`? From what I see it is called before `PrePredicateFn` so there is probably something that you identified as a gap in the implementation?
> 
> Another question, what is the problem with job allocation above parent's queue quota? If the job is preemptible it can be allowed.

Hi @romanbaron 👋

Thank you for your questions! Let me address them:

1. **Gap with `IsJobOverQueueCapacity`**:
   - `IsJobOverQueueCapacity` only checks the immediate queue's capacity
   - Our new `checkParentQueueQuotas` function checks ALL parent queues up the hierarchy
   - This is important because a job might fit in its immediate queue but exceed parent queue quotas
   - By checking in `PrePredicateFn`, we fail fast and avoid unnecessary scheduling attempts

2. **Preemptible Jobs and Parent Queue Quotas**:
   - You're absolutely right! Preemptible jobs should be allowed to exceed parent queue quotas
   - I've updated the implementation to handle this:
     - Added a check for `PriorityTrainNumber` at the start of `checkParentQueueQuotas`
     - Preemptible jobs now skip quota checks entirely
     - Non-preemptible jobs still maintain strict quota enforcement
   - This change is reflected in the test cases:
     - `preemptible_job_can_exceed_parent_queue_GPU_quota`
     - `non-preemptible_job_cannot_exceed_parent_queue_GPU_quota`

The changes ensure that:
- Preemptible jobs can utilize resources beyond parent queue quotas
- Non-preemptible jobs maintain strict quota enforcement
- We still get early validation for jobs that would exceed quotas

enoodle commented on pkg/scheduler/plugins/proportion/capacity_policy/capacity_policy.go:
```diff
@@ -83,3 +113,131 @@ func getRequiredQuota(tasksToAllocate []*pod_info.PodInfo) *podgroup_info.JobReq
+		return nil // Can't check quota without queue info
+	}
+
+	// Only check parent queues, not the job's direct queue
+	currentQueueID := queue.ParentQueue
+
+	for currentQueueID != "" {
+		parentQueue, found := ssn.Queues[currentQueueID]
+		if !found {
+			break
+		}
```
This together with line 239 can be one line of `for` with initialization, check and step

enoodle commented on pkg/scheduler/plugins/proportion/capacity_policy/capacity_policy.go:
```diff
@@ -83,3 +113,131 @@ func getRequiredQuota(tasksToAllocate []*pod_info.PodInfo) *podgroup_info.JobReq
+			}
+		}
+
+		// Check GPU quota
```
this part seems duplicated 3 times, there are functions to avoid it in the code base. you can loop on the resource type and the . look at proportion/proportion.go for example.

enoodle commented on pkg/scheduler/plugins/proportion/capacity_policy/capacity_policy.go:
```diff
@@ -83,3 +113,131 @@ func getRequiredQuota(tasksToAllocate []*pod_info.PodInfo) *podgroup_info.JobReq
+		// Check GPU quota
+		if parentQueue.Resources.GPU.Quota > 0 && jobResources.GPUs() > float64(parentQueue.Resources.GPU.Quota) {
+			errorMsg := fmt.Sprintf(
+				"parent queue '%s' quota has reached the allowable limit of GPUs. "+
```
A queue can go over the quota, but not the limit. I would change the whole thing to address the limit.

enoodle commented on pkg/scheduler/plugins/proportion/capacity_policy/capacity_policy.go:
```diff
@@ -83,3 +113,131 @@ func getRequiredQuota(tasksToAllocate []*pod_info.PodInfo) *podgroup_info.JobReq
+// queue quotas, while non-preemptible jobs must strictly adhere to quotas.
+func (cp *CapacityPolicy) checkParentQueueQuotas(job *podgroup_info.PodGroupInfo, ssn *framework.Session) error {
+	// Skip quota checks for preemptible jobs
+	if job.Priority == constants.PriorityTrainNumber {
```
We aim to separate the job priority and preemptibility soon, and the Train priority is not the only one that is currently preemptible.
We do have several scalars to define the resource of a queue: quota, limit and over quota weight. We should look at limit here and not quota and then it will be correct for all jobs.

enoodle reviewed (commented):
Hi,
I left a few comments, mostly about style.

I do have a question about the logic: how will this handle elastic jobs who doesn't have to schedule all pods to run? The job has a minimal number of pods that it needs to run, and you don't know how many will actually fit in the cluster / queue limits.

enoodle commented on pkg/scheduler/plugins/proportion/capacity_policy/capacity_policy.go:
```diff
@@ -83,3 +113,131 @@ func getRequiredQuota(tasksToAllocate []*pod_info.PodInfo) *podgroup_info.JobReq
+	// Register early quota checks
+	ssn.AddPrePredicateFn(func(task *pod_info.PodInfo, job *podgroup_info.PodGroupInfo) error {
+		// Only check for the first pending pod to avoid duplicate checks
+		firstPending := getFirstPendingPod(job)
```
Isn't the fact that the job needs scheduling (and then predicates are running on it) is not enough to understand that it has pending pods?

Commit:
Add parent queue limit and quota checking with elastic job support

shchennu commented:
> Hi, I left a few comments, mostly about style.
> 
> I do have a question about the logic: how will this handle elastic jobs who doesn't have to schedule all pods to run? The job has a minimal number of pods that it needs to run, and you don't know how many will actually fit in the cluster / queue limits.

Thanks for the review! I've addressed all the comments:
- Added support for elastic jobs with minimum resource requirements
- Improved code style by removing redundant comments and combining loops
- Now checking both limits and quotas, using the more restrictive value
- Separated job priority from preemptibility
- Added test coverage

All tests are passing. Let me know if you'd like any further improvements!

enoodle commented:
> 1. **Gap with `IsJobOverQueueCapacity`**:
>    
>    * `IsJobOverQueueCapacity` only checks the immediate queue's capacity
>    * Our new `checkParentQueueQuotas` function checks ALL parent queues up the hierarchy
>    * This is important because a job might fit in its immediate queue but exceed parent queue quotas
>    * By checking in `PrePredicateFn`, we fail fast and avoid unnecessary scheduling attempts


@shchennu I looked at `IsJobOverQueueCapacity` and I don't agree with your observation - it is checking all the parent queues, so this addition is not needed.

romanbaron commented:
@shchennu - all hierarchy levels are validated in these two places:
[resultsOverLimit](https://github.com/NVIDIA/KAI-Scheduler/blob/2a0bacdbba1444b72236373c84fc8be742e57b63/pkg/scheduler/plugins/proportion/capacity_policy/max_allowed_check.go#L19) - checks that workload scheduling won't cause any of the queues (from leaf to top) to go over limit.
[resultsWithNonPreemptibleOverQuota](https://github.com/NVIDIA/KAI-Scheduler/blob/2a0bacdbba1444b72236373c84fc8be742e57b63/pkg/scheduler/plugins/proportion/capacity_policy/quota_check.go#L34) - checks that non-preemptible workload scheduling won't cause any of the queues (from leaf to top) to go over quota.

IsJobOverQueueCapacity calls these two functions here:
https://github.com/NVIDIA/KAI-Scheduler/blob/2a0bacdbba1444b72236373c84fc8be742e57b63/pkg/scheduler/plugins/proportion/capacity_policy/capacity_policy.go#L35

Doesn't it cover it?

romanbaron commented:
@shchennu - can you please update on your thoughts regarding the questions above?
